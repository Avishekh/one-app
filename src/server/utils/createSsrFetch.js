/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
import http from 'http';
import https from 'https';

const baseAgentOptions = {
  keepAlive: true,
  keepAliveMsecs: 1000,
  maxFreeSockets: 5,
  maxSockets: 100,
};

export function enhancedSsrFetch() {
  const httpAgent = new http.Agent(baseAgentOptions);

  const httpsAgent = new https.Agent(baseAgentOptions);

  function selectAgent(urlObject) {
    if (urlObject.protocol === 'http:') {
      return httpAgent;
    }
    return httpsAgent;
  }

  return (fetch) => (url, opts) => fetch(url, { ...opts, agent: selectAgent(url) });
}

const cache = {
  createSsrFetch: enhancedSsrFetch,
};

export const setCreateSsrFetch = (newFetch = enhancedSsrFetch) => {
  cache.createSsrFetch = newFetch;
};

const createSsrFetch = (opts = {}) => cache.createSsrFetch(opts);

export default createSsrFetch;
