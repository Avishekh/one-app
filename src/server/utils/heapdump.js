/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// NODE_HEAPDUMP_OPTIONS=nosignal by default in server/config/env/runTime
let heapdump;
try {
  // binary might not be compatible for non-docker distributions
  heapdump = require('heapdump'); // eslint-disable-line global-require, import/no-extraneous-dependencies, import/no-unresolved
} catch (err) {
  console.warn(
    'unable to setup writing heapdumps',
    process.env.NODE_ENV === 'development' ? err.message : err
  );
}

if (heapdump && process.env.NODE_HEAPDUMP_OPTIONS === 'nosignal') {
  process.on('SIGUSR2', () => {
    const targetFilename = `/tmp/heapdump-${process.pid}-${Date.now()}.heapsnapshot`;
    console.warn(`about to write a heapdump to ${targetFilename}`);
    heapdump.writeSnapshot(targetFilename, (err, writtenFilename) => {
      if (err) {
        console.error(`unable to write heapdump ${writtenFilename}`, err);
      } else {
        console.warn(`wrote heapdump out to ${writtenFilename}`);
      }
    });
  });
}
