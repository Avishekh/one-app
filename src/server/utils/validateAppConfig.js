/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Joi from '@hapi/joi';

const webManifestSchema = Joi.object().keys({
  name: Joi.string().required(),
  short_name: Joi.string(),
  lang: Joi.string(),
  scope: Joi.string().pattern(/^\//),
  start_url: Joi.string().pattern(/^\//),
  theme_color: Joi.alternatives(
    Joi.string(),
    Joi.string().hex()
  ),
  background_color: Joi.alternatives(
    Joi.string(),
    Joi.string().hex()
  ),
  categories: Joi.array().items(Joi.string()),
  iarc_rating_id: Joi.boolean(),
  prefer_related_applications: Joi.boolean(),
  related_applications: Joi.array().items(Joi.object().keys({
    platform: Joi.string(),
    url: Joi.string(),
    id: Joi.string(),
  })),
  orientation: Joi.string().valid(
    'any',
    'natural',
    'landscape',
    'landscape-primary',
    'landscape-secondary',
    'portrait',
    'portrait-primary',
    'portrait-secondary'
  ),
  display: Joi.string().valid(
    'fullscreen',
    'standalone',
    'minimal-ui',
    'browser'
  ),
  dir: Joi.string().valid(
    'auto',
    'ltr',
    'rtl'
  ),
  icons: Joi.array().items(Joi.object().keys({
    src: Joi.string().uri(),
    sizes: Joi.string(),
    type: Joi.string(),
    purpose: Joi.string().valid(
      'any',
      'maskable',
      'badge'
    ),
  })),
  screenshots: Joi.array().items(Joi.object().keys({
    src: Joi.string().uri(),
    sizes: Joi.string(),
    type: Joi.string(),
  })),
});

const pwaSchema = Joi.object().keys({
  serviceWorker: Joi.boolean(),
  recoveryMode: Joi.boolean(),
  escapeHatch: Joi.boolean(),
  scope: Joi.string().pattern(/^\//),
  webManifest: Joi.alternatives().try(webManifestSchema, Joi.function()),
});

const safeRequestSchema = Joi.object().keys({
  headers: Joi.array().items(Joi.string()),
  cookies: Joi.array().items(Joi.string()),
});

const stateConfigKeySchema = Joi.object().unknown().keys({
  server: Joi.string().required().allow(null),
  client: Joi.string().required().allow(null),
});

const provideStateConfigSchema = Joi.object().pattern(
  Joi.string(),
  Joi.object().concat(stateConfigKeySchema)
);

// TODO: add custom messages for fields
// TODO: validate app compatibility via semver

const rootModuleSchema = Joi.object().keys({
  providedExternals: Joi.array().items(Joi.string()),
  provideStateConfig: provideStateConfigSchema,
  csp: Joi.string().required(),
  corsOrigins: Joi.array().items(Joi.string().uri({
    scheme: [
      /https?/,
    ],
  })),
  configureRequestLog: Joi.function(),
  extendSafeRequestRestrictedAttributes: safeRequestSchema,
  pwa: pwaSchema,
  createSsrFetch: Joi.function(),
  eventLoopDelayThreshold: Joi.number().allow(Infinity),
  appCompatibility: Joi.string(),
});

const childModuleSchema = Joi.object().keys({
  requiredExternals: Joi.array().items(Joi.string()),
  validateStateConfig: Joi.object().pattern(
    Joi.string(),
    Joi.object().concat(stateConfigKeySchema)
  ),
  requiredSafeRequestRestrictedAttributes: safeRequestSchema,
  appCompatibility: Joi.string(),
});

function validateSchema(schema, validationTarget) {
  const { error, value } = schema.validate(validationTarget, { abortEarly: false, render: false });
  if (error) throw error;
  return value;
}

export function validateWebAppManifest(webManifest) {
  return validateSchema(webManifestSchema, webManifest);
}

// TODO: validate provided/required state config
// TODO: validate provided/required safe request restricted attributes

export function validateRootModuleAppConfig(appConfig) {
  return validateSchema(rootModuleSchema, appConfig);
}

export function validateChildModuleAppConfig(appConfig) {
  // TODO: move validateStateConfig to here
  return validateSchema(childModuleSchema, appConfig);
}
