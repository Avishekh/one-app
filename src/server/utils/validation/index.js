/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Joi from '@hapi/joi';

import { provideStateConfigSchema, validateStateConfigSchema } from './stateConfig';
import { providedExternalSchema, requiredExternalSchema } from './externals';
import { safeRequestSchema } from './safeRequest';
import { semverExtension } from './semver';
import { pwaSchema } from './pwa';

export const rootModuleSchema = Joi.object().keys({
  providedExternals: providedExternalSchema,
  provideStateConfig: provideStateConfigSchema,
  csp: Joi.string().required().messages({
    'any.required': '"csp" must provide a valid content security policy in the Root module',
  }),
  corsOrigins: Joi.array().items(
    Joi.alternatives().try(
      Joi.object().instance(RegExp),
      Joi.string().hostname(),
      Joi.string().pattern(/(?:https?)?[a-z.-]/)
    )
  ),
  configureRequestLog: Joi.function(),
  extendSafeRequestRestrictedAttributes: safeRequestSchema,
  pwa: pwaSchema,
  createSsrFetch: Joi.function(),
  eventLoopDelayThreshold: Joi.number().allow(Infinity),
  appCompatibility: semverExtension.semver().valid(),
});

export const childModuleSchema = Joi.object().keys({
  providedExternals: Joi.any().custom((value, helpers) => helpers.warn('any.exists')).messages({
    'any.exists': 'Module {{$moduleName}} attempted to provide externals. Only the root module can provide externals.',
  }),
  requiredExternals: requiredExternalSchema,
  validateStateConfig: validateStateConfigSchema,
  requiredSafeRequestRestrictedAttributes: safeRequestSchema,
  appCompatibility: semverExtension.semver().valid(),
});


function validateSchema(schema, validationTarget, options) {
  const { error, value, warning } = schema.validate(
    validationTarget,
    { ...options, abortEarly: false }
  );
  if (error) throw error;
  if (warning) {
    console.warn(warning);
  }
  return value;
}

export function validateRootModuleAppConfig(appConfig, context) {
  return validateSchema(rootModuleSchema, appConfig, { context });
}

export function validateChildModuleAppConfig(appConfig, context) {
  return validateSchema(childModuleSchema, appConfig, { context });
}
