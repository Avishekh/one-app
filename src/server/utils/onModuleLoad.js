/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { META_DATA_KEY } from '@americanexpress/one-app-bundler';

import { setStateConfig, getClientStateConfig, getServerStateConfig } from './stateConfig';
import { setCorsOrigins } from '../middleware/conditionallyAllowCors';
import readJsonFile from './readJsonFile';
import { extendRestrictedAttributesAllowList, validateSafeRequestRestrictedAttributes } from './safeRequest';
import { setConfigureRequestLog } from './logging/serverMiddleware';
import { setCreateSsrFetch } from './createSsrFetch';
import { setEventLoopDelayThreshold } from './createCircuitBreaker';
import { configurePWA } from '../middleware/pwa';
import { updateCSP } from '../middleware/csp';
import { validateRootModuleAppConfig, validateChildModuleAppConfig } from './validation';

// Trim build hash
const { buildVersion } = readJsonFile('../../../.build-meta.json');
const appVersion = buildVersion.slice(0, -9);
let modulesUsingExternals = new Set();

const registerModuleUsingExternals = (moduleName) => {
  modulesUsingExternals = modulesUsingExternals.add(moduleName);
};

const clearModulesUsingExternals = () => {
  modulesUsingExternals = modulesUsingExternals.clear();
};

export const getModulesUsingExternals = () => [...modulesUsingExternals || []];

export const setModulesUsingExternals = (moduleNames) => {
  modulesUsingExternals = new Set(moduleNames);
};

const logModuleLoad = (moduleName, moduleVersion) => {
  console.info(`Loaded module ${moduleName}@${moduleVersion}`);
};

function validateConfig(configValidators, config) {
  Object.entries(configValidators)
    .forEach(([configKey, {
      client: { validate: validateClient = () => 0 } = {},
      server: { validate: validateServer = () => 0 } = {},
    }]) => {
      validateClient(config.client[configKey]);
      validateServer(config.server[configKey]);
    });
}

export const CONFIGURATION_KEY = 'appConfig';

export default function onModuleLoad({
  module,
  moduleName,
}) {
  const {
    [CONFIGURATION_KEY]: moduleConfig = {},
    [META_DATA_KEY]: metaData = {},
  } = module;

  const serverStateConfig = getServerStateConfig();
  const clientStateConfig = getClientStateConfig();
  const validationContext = {
    moduleName,
    moduleVersion: metaData.version,
    appVersion,
    serverStateConfig,
    clientStateConfig,
  };

  if (moduleName === serverStateConfig.rootModuleName) {
    const {
      provideStateConfig,
      csp,
      corsOrigins,
      configureRequestLog,
      extendSafeRequestRestrictedAttributes,
      createSsrFetch,
      eventLoopDelayThreshold,
      pwa,
    } = validateRootModuleAppConfig(moduleConfig, validationContext);

    clearModulesUsingExternals();
    if (provideStateConfig) {
      setStateConfig(provideStateConfig);
    }
    updateCSP(csp);
    setCorsOrigins(corsOrigins);
    extendRestrictedAttributesAllowList(extendSafeRequestRestrictedAttributes);
    setConfigureRequestLog(configureRequestLog);
    setCreateSsrFetch(createSsrFetch);
    setEventLoopDelayThreshold(eventLoopDelayThreshold);
    configurePWA(pwa);
  } else {
    const RootModule = global.getTenantRootModule();
    const { providedExternals } = RootModule[CONFIGURATION_KEY];
    validationContext.providedExternals = { ...providedExternals };

    const {
      requiredExternals,
      validateStateConfig,
      requiredSafeRequestRestrictedAttributes,
    } = validateChildModuleAppConfig(moduleConfig, validationContext);

    if (validateStateConfig) {
      validateConfig(validateStateConfig, {
        server: serverStateConfig,
        client: clientStateConfig,
      });
    }

    if (requiredExternals) registerModuleUsingExternals(moduleName);

    validateSafeRequestRestrictedAttributes(requiredSafeRequestRestrictedAttributes);
  }

  logModuleLoad(moduleName, metaData.version);
}
