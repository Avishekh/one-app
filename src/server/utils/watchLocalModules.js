/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// This file is only used in development so importing devDeps is not an issue
/* eslint "import/no-extraneous-dependencies": ["error", {"devDependencies": true}] */

import fs from 'fs';
import path from 'path';
import chokidar from 'chokidar';
import loadModule from 'holocron/loadModule.node';
import {
  getModules,
  getModuleMap,
  resetModuleRegistry,
} from 'holocron/moduleRegistry';
import onModuleLoad from './onModuleLoad';

export default function watchLocalModules() {
  const staticsDirectoryPath = path.resolve(__dirname, '../../../static');
  const moduleDirectory = path.resolve(staticsDirectoryPath, 'modules');
  const moduleMapPath = path.resolve(staticsDirectoryPath, 'module-map.json');
  const watcher = chokidar.watch(moduleDirectory);

  watcher.on('change', async (changedPath) => {
    if (!changedPath.endsWith('.node.js')) return;

    const match = changedPath.substring(moduleDirectory.length).match(/\/([^/]+)\/([^/]+)/);
    if (!match) return;

    const [, moduleNameChangeDetectedIn] = match;

    const moduleMap = JSON.parse(fs.readFileSync(moduleMapPath, 'utf8'));

    const module = await loadModule(
      moduleNameChangeDetectedIn,
      moduleMap.modules[moduleNameChangeDetectedIn],
      onModuleLoad
    );

    const modules = getModules().set(moduleNameChangeDetectedIn, module);
    resetModuleRegistry(modules, getModuleMap());
  });
}
