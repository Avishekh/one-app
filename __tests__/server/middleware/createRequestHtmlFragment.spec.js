/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'url';
import { browserHistory } from '@americanexpress/one-app-router';
import { Map as iMap, fromJS } from 'immutable';
import match from '../../../src/universal/utils/matchPromisified';

import * as reactRendering from '../../../src/server/utils/reactRendering';

jest.mock('../../../src/universal/utils/matchPromisified');

jest.mock('holocron', () => ({
  composeModules: jest.fn(() => 'composeModules'),
}));

const renderForStringSpy = jest.spyOn(reactRendering, 'renderForString');
const renderForStaticMarkupSpy = jest.spyOn(reactRendering, 'renderForStaticMarkup');

describe('createRequestHtmlFragment', () => {
  jest.spyOn(console, 'error').mockImplementation(() => {});

  let req;
  let res;
  let next;
  let createRoutes;
  const dispatch = jest.fn((x) => x);
  const getState = jest.fn(() => fromJS({
    rendering: iMap({}),
  }));

  beforeAll(() => {
    match.mockImplementation(({ routes, location }) => Promise.resolve({
      redirectLocation: undefined,
      renderProps: {
        routes,
        components: [() => 'hi'],
        // history: browserHistory,
        location: url.parse(location),
        router: browserHistory,
        params: {
          hi: 'there?',
        },
      },
    }));
  });

  beforeEach(() => {
    req = jest.fn();
    req.headers = {};

    res = jest.fn();
    res.status = jest.fn(() => res);
    res.sendStatus = jest.fn(() => res);
    res.redirect = jest.fn(() => res);
    res.end = jest.fn(() => res);
    req.url = 'http://example.com/request';
    req.store = { dispatch, getState };

    next = jest.fn();

    createRoutes = jest.fn(() => [{ path: '/', moduleName: 'root' }]);

    renderForStringSpy.mockClear();
    renderForStaticMarkupSpy.mockClear();
  });

  it('returns a function', () => {
    const createRequestHtmlFragment = require(
      '../../../src/server/middleware/createRequestHtmlFragment'
    ).default;
    const middleware = createRequestHtmlFragment({ createRoutes });
    expect(middleware).toBeInstanceOf(Function);
  });

  it('should preload data for matched route components', () => {
    const createRequestHtmlFragment = require(
      '../../../src/server/middleware/createRequestHtmlFragment'
    ).default;
    const middleware = createRequestHtmlFragment({ createRoutes });
    return middleware(req, res, next)
      .then(() => {
        const { composeModules } = require('holocron');
        expect(composeModules).toHaveBeenCalled();
        expect(composeModules.mock.calls[0][0]).toMatchSnapshot();
        expect(dispatch).toHaveBeenCalledWith('composeModules');
      });
  });

  it('should add app HTML to the request object', () => {
    const createRequestHtmlFragment = require(
      '../../../src/server/middleware/createRequestHtmlFragment'
    ).default;
    const middleware = createRequestHtmlFragment({ createRoutes });
    return middleware(req, res, next)
      .then(() => {
        expect(next).toHaveBeenCalled();
        expect(createRoutes).toHaveBeenCalledWith(req.store);
        expect(renderForStringSpy).toHaveBeenCalled();
        expect(renderForStaticMarkupSpy).not.toHaveBeenCalled();
        expect(typeof req.appHtml).toBe('string');
      });
  });

  it('should add app HTML as static markup to the request object when scripts are disabled', () => {
    const createRequestHtmlFragment = require(
      '../../../src/server/middleware/createRequestHtmlFragment'
    ).default;
    const middleware = createRequestHtmlFragment({ createRoutes });
    getState.mockImplementationOnce(() => fromJS({
      rendering: iMap({ disableScripts: true }),
    }));
    return middleware(req, res, next)
      .then(() => {
        expect(next).toHaveBeenCalled();
        expect(createRoutes).toHaveBeenCalledWith(req.store);
        expect(renderForStringSpy).not.toHaveBeenCalled();
        expect(renderForStaticMarkupSpy).toHaveBeenCalled();
        expect(typeof req.appHtml).toBe('string');
      });
  });

  it('should set the custom HTTP status', () => {
    const createRequestHtmlFragment = require(
      '../../../src/server/middleware/createRequestHtmlFragment'
    ).default;

    createRoutes = jest.fn(() => [{ path: '/', httpStatus: 400 }]);
    const middleware = createRequestHtmlFragment({ createRoutes });
    return middleware(req, res, next)
      .then(() => {
        expect(next).toHaveBeenCalled();
        expect(createRoutes).toHaveBeenCalledWith(req.store);
        expect(res.status).toHaveBeenCalledWith(400);
        expect(typeof req.appHtml).toBe('string');
      });
  });

  it('does not generate HTML when no route is matched', () => {
    const createRequestHtmlFragment = require(
      '../../../src/server/middleware/createRequestHtmlFragment'
    ).default;

    match.mockImplementationOnce(() => ({
      redirectLocation: undefined,
      // omit renderProps
    }));

    const middleware = createRequestHtmlFragment({ createRoutes });
    // eslint-disable-next-line no-console
    console.error = jest.fn();
    return middleware(req, res, next)
      .then(() => {
        // eslint-disable-next-line no-console
        expect(console.error).toHaveBeenCalled();
        expect(next).toHaveBeenCalled();
        expect(res.sendStatus).toHaveBeenCalledWith(404);
        expect(createRoutes).toHaveBeenCalledWith(req.store);
        expect(req.appHtml).toBe(undefined);
      });
  });

  it('redirects when a relative redirect route is matched', () => {
    const createRequestHtmlFragment = require(
      '../../../src/server/middleware/createRequestHtmlFragment'
    ).default;

    match.mockImplementationOnce(() => ({
      redirectLocation: {
        pathname: '/redirect',
        search: '',
      },
    }));

    const middleware = createRequestHtmlFragment({ createRoutes });
    return middleware(req, res, next)
      .then(() => {
        expect(createRoutes).toHaveBeenCalledWith(req.store);
        expect(res.redirect).toHaveBeenCalledWith(302, '/redirect');
        expect(req.appHtml).toBe(undefined);
      });
  });

  it('redirects when an absolute redirect route is matched', () => {
    const createRequestHtmlFragment = require(
      '../../../src/server/middleware/createRequestHtmlFragment'
    ).default;

    match.mockImplementationOnce(() => ({
      redirectLocation: {
        state: url.parse('https://example.com/redirect'),
      },
    }));

    const middleware = createRequestHtmlFragment({ createRoutes });
    return middleware(req, res, next)
      .then(() => {
        expect(createRoutes).toHaveBeenCalledWith(req.store);
        expect(res.redirect).toHaveBeenCalledWith(302, 'https://example.com/redirect');
        expect(req.appHtml).toBe(undefined);
      });
  });

  it('should catch any errors and call the next middleware', () => {
    const createRequestHtmlFragment = require(
      '../../../src/server/middleware/createRequestHtmlFragment'
    ).default;

    delete req.store;
    const middleware = createRequestHtmlFragment({ createRoutes: null });
    /* eslint-disable no-console */
    console.error = jest.fn();
    middleware(req, res, next);
    expect(console.error).toHaveBeenCalled();
    expect(console.error.mock.calls[0]).toMatchSnapshot();
    expect(next).toHaveBeenCalled();
    /* eslint-enable no-console */
  });
});
