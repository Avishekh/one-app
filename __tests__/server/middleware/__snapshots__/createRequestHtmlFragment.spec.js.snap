// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRequestHtmlFragment should catch any errors and call the next middleware 1`] = `
Array [
  "error creating request HTML fragment for http://example.com/request",
  [TypeError: Cannot read property 'dispatch' of undefined],
]
`;

exports[`createRequestHtmlFragment should preload data for matched route components 1`] = `
Array [
  Object {
    "name": "root",
    "props": Object {
      "location": Url {
        "auth": null,
        "hash": null,
        "host": "example.com",
        "hostname": "example.com",
        "href": "http://example.com/request",
        "path": "/request",
        "pathname": "/request",
        "port": null,
        "protocol": "http:",
        "query": null,
        "search": null,
        "slashes": true,
      },
      "params": Object {
        "hi": "there?",
      },
      "route": Object {
        "moduleName": "root",
        "path": "/",
      },
      "router": Object {
        "createHref": [Function],
        "createKey": [Function],
        "createLocation": [Function],
        "createPath": [Function],
        "getCurrentLocation": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "listenBefore": [Function],
        "push": [Function],
        "replace": [Function],
        "transitionTo": [Function],
      },
      "routes": Array [
        Object {
          "moduleName": "root",
          "path": "/",
        },
      ],
    },
  },
]
`;
